conf/evolutions/default/221.sql

```diff
@@ -0,0 +1,12 @@
+# --- !Ups
+CREATE TABLE pano_history (
+   pano_id TEXT NOT NULL,
+   visited_timestamp TIMESTAMP,
```

Comment:
Ohhh I see, it's because that's the timestamp of only the current pano... What if instead, this became a new column in the `gsv_data` table, called something like `pano_history_saved`? It would still have to be nullable, but at least it would be nullable because we haven't recorded the history for all panos, instead nulls being an indication of there being different types of rows in the same table.

On that front, it might make more sense to make the foreign key point to the `gsv_data` table as well? Just to make it clear that this is an auxiliary table to `gsv_data`. And then do we even bother to store the current pano in the `pano_history` table?

Reply:
Sure, I can definitely do that! This does mean the procedure when updating a new current pano at some location will be different but that shouldn't be too bad. One question I did have though is are we guaranteed that every pano we see in the validate page is already inside the GSVDataTable? I do kinda need to know that since I'll have to edit that some row's pano_history_saved field in that table every time I insert a new pano history, and that validation route does not add stuff to that table. I assumed the route that handles all the label task submissions handles adding every single pano in label (and that eventually shows up in validate) to that table, but I do just wanna confirm!

Reply:
Also, the GSVDataTable has a field called last_viewed which gets updated whenever it gets visited. Can I just use that field instead of creating a new pano_history_saved field? I can't really think of a case where we'd need to make the new one

Reply:
>are we guaranteed that every pano we see in the validate page is already inside the GSVDataTable

This should be the case, yes! When querying for labels for the Validate page, we do an inner join with the `gsv_data` table.

>GSVDataTable has a field called last_viewed which gets updated whenever it gets visited. Can I just update that field whenever a panorama in that table is viewed by someone in label or validate as opposed to making the new pano_history_saved field

That won't work because the `last_viewed` field is doing something slightly different. This column can get updated after we check if imagery exists there using the free metadata API, which doesn't necessarily require a user visiting the pano on Explore or Validate. So unfortunately these need to be separate!

Reply:
Sounds good, I'll go ahead and add it! One more thing for code simplicity - that last_viewed field is just the timestamp of a new Date object (which is data.timestamp in the backend) created when the data is submitted to the backend. For the visited_timestamp field that I had in the PanoHistory table, I was getting the exact time that the user visited the pano by editing the Panorama object in the frontend. For simplicity's sake is it okay if I get rid of that and just use the same new date object that gets passed to the backend for the pano_history_saved field?

Reply:
Yeah that's fine! The timestamp doesn't need to be that accurate